File: app/src/main/java/com/example/mvvm_artichecture_sample/base/BaseFragment.java
package com.example.mvvm_artichecture_sample.base;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import io.reactivex.disposables.Disposable;

public abstract class BaseFragment<V extends BaseViewModel> extends Fragment {

    protected View currView;
    protected String currLayoutName;
    protected Context mContext;
    protected boolean isActive;
    private String mClassName = null;
    protected V mViewModel;

    protected abstract V getViewModel();

    /**
     * @return layout resource id
     */
    protected abstract
    @LayoutRes
    int getLayoutId();


    /**
     * This method called after initialize view
     */
    protected abstract void setupView();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mViewModel = getViewModel();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        if (currView == null) {
            setLayoutView(getLayoutId(), inflater, container, "");
            setupView();
        }
        return currView;
    }


    public void setLayoutView(int layoutId, LayoutInflater inflater, ViewGroup container) {
        currView = inflater.inflate(layoutId, container, false);

        currLayoutName = getResources().getResourceEntryName(layoutId);
        mClassName = currLayoutName;
        mContext = getContext();
    }

    public void setLayoutView(int layoutId, LayoutInflater inflater, ViewGroup container,
            String className) {
        setLayoutView(layoutId, inflater, container);

        mClassName = className;
    }

    public void addDisposable(Disposable disposable) {
        mViewModel.getCompositeDisposable().add(disposable);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        getViewLifecycleOwnerLiveData().removeObservers(this);

    }

    protected void back() {
        if (getActivity() != null) {
            getActivity().onBackPressed();
        }
    }

}

File: app/src/main/java/com/example/mvvm_artichecture_sample/base/BaseFragment.kt
package com.example.mvvm_artichecture_sample.base

import android.app.Application
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment

import io.reactivex.disposables.Disposable

abstract class BaseFragment<V : BaseViewModel<*>> : Fragment() {

    protected var currView: View? = null
    protected var currLayoutName: String?=null
    protected var activityContext: Context? = null
    protected var isActive: Boolean = false
    private var mClassName: String? = null
    public var mViewModel: V?=null

    protected abstract val viewModel: V

    /**
     * @return layout resource id
     */
    @get:LayoutRes
    protected abstract val layoutId: Int

    protected val application: Application
        get() = if (activity == null)
            throw IllegalStateException("Your activity/fragment is not yet attached to " + "Application.")
        else
            activity!!.application


    /**
     * This method called after initialize view
     */
    protected abstract fun setupView()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mViewModel = viewModel
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (currView == null) {
            setLayoutView(layoutId, inflater, container, "")
            setupView()
        }
        return currView
    }


    fun setLayoutView(layoutId: Int, inflater: LayoutInflater, container: ViewGroup?) {
        currView = inflater.inflate(layoutId, container, false)

        currLayoutName = resources.getResourceEntryName(layoutId)
        mClassName = currLayoutName
        activityContext = context
    }

    fun setLayoutView(layoutId: Int, inflater: LayoutInflater, container: ViewGroup?,
                      className: String) {
        setLayoutView(layoutId, inflater, container)

        mClassName = className
    }

    fun addDisposable(disposable: Disposable) {
        mViewModel!!.compositeDisposable.add(disposable)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        viewLifecycleOwnerLiveData.removeObservers(this)

    }

    protected fun back() {
        if (activity != null) {
            activity!!.onBackPressed()
        }
    }
}
